sycl/plugins/opencl/pi_opencl.cpp

```diff
@@ -126,6 +127,80 @@ static pi_result getExtFuncFromContext(pi_context context, T *fptr) {
   return cast<pi_result>(ret_err);
 }
 
+// This is a temporary workaround for a clMemBlockingFreeINTEL bug in OpenCL CPU
+// runtime. Chooses clMemBlockingFreeINTEL for GPU and clMemFreeINTEL for other
+// devices.
+// TODO remove this workaround once the OpenCL CPU RT version with the fix is
+// uplifted
+static pi_result getUSMFreeFunc(pi_context context,
+                                clMemBlockingFreeINTEL_fn *fptr) {
+  static_assert(
+      std::is_same<clMemBlockingFreeINTEL_fn, clMemFreeINTEL_fn>::value);
+  thread_local static std::map<pi_context, clMemBlockingFreeINTEL_fn> FuncPtrs;
+
+  // if cached, return cached FuncPtr
+  if (auto F = FuncPtrs[context]) {
+    // if cached that extension is not available return nullptr and
+    // PI_INVALID_VALUE
+    *fptr = F;
+    return F ? PI_SUCCESS : PI_INVALID_VALUE;
+  }
+
+  cl_uint deviceCount;
+  cl_int ret_err =
+      clGetContextInfo(cast<cl_context>(context), CL_CONTEXT_NUM_DEVICES,
```

Comment:
why are you not re-using `getExtFuncFromContext`?

Reply:
I'd like to, but for this workaround I need to replicate `getExtFuncFromContext` behaviour (i.e. retrieving and caching extension function pointers) with the difference that the chosen function is dependent on device type (`clMemBlockingFreeINTEL` for GPU, `clMemFreeINTEL` for the rest). When this workaround is no longer needed, `getExtFuncFromContext` will be used instead.

Reply:
why can't you query/cache both versions (blocking and non-blocking) and use the one as needed under some condition?

Reply:
@smaslov-intel I suppose I could query/cache separately which contexts are GPU and which are not, and then call the corresponding instantiation of `getExtFuncFromContext` which will then query/cache the chosen function for that context. But since each context will always call only one of USM free functions, I wanted to skip the middleman and cache the needed USM free function directly.
Do you think the other approach would be better here? I don't feel particularly strong one way or another.

Reply:
Why is it not as simple as 

```
if (GPU)
  return getExtFuncFromContext<clMemBlockingFreeINTEL> (Context);
else 
  return getExtFuncFromContext<clMemFreeINTEL> (Context);
```

Reply:
Because I'd like to avoid both a) calling clGetDeviceInfo each time USM free is called; b) introducing extra caching of which contexts are GPU and which are not for this workaround (which is not information that's tracked in OpenCL plugin currently). That's why I initially chose c) going with an altered version of `getExtFuncFromContext` that takes device type into consideration to cache the correct function.
Although, again, if you're not convinced by that and think that option A or B would be better here (I'm not sure which one you're suggesting to go with), I'm willing to switch to that.

Reply:
Switched to the option A as discussed.